Fix #2644 traefik chart duplicate value

raykrueger opened the pull request:
The logic in place states that if ssl is **NOT** enabled in traefik,
then have the kube service listen on 443 and send to 80. There is
already a port/targetPort setup for 80.

This results in the error described in issue #2644

    Error: release traefik failed: Service "traefik-traefik" is invalid:
spec.ports[1].targetPort: Duplicate value: intstr.IntOrString{Type:0,
IntVal:80, StrVal:""}

My fix is when ssl is not enabled, to add a second traefik listening
port at 84. Which means the service is setup with the following port
mappings:

    80  -> 80
    443 -> 84

The ssl.enabled logic in this chart does not seem to refer to the
service, only traefik itself. My solution is a little copy pasta in that
I add the second http listener, but then I'm not changing/breaking the
ssl.enabled logic.

raykrueger committed:
Fix #2644 traefik chart duplicate value

The logic in place states that if ssl is **NOT** enabled in traefik,
then have the kube service listen on 443 and send to 80. There is
already a port/targetPort setup for 80.

This results in the error described in issue #2644

    Error: release traefik failed: Service "traefik-traefik" is invalid:
spec.ports[1].targetPort: Duplicate value: intstr.IntOrString{Type:0,
IntVal:80, StrVal:""}

My fix is when ssl is not enabled, to add a second traefik listening
port at 84. Which means the service is setup with the following port
mappings:

    80  -> 80
    443 -> 84

The ssl.enabled logic in this chart does not seem to refer to the
service, only traefik itself. My solution is a little copy pasta in that
I add the second http listener, but then I'm not changing/breaking the
ssl.enabled logic.

raykrueger committed:
Bump chart patch version

(oops)

raykrueger committed:
Match the endpoint names correctly

krancour commented:
I realize that k8s 1.8 broke this, but the original behavior was not a mistake. It was quite deliberate because (even though I discourage it personally) there are many, many people who terminate SSL at the load balancer. They want the load balancer to listen on 443 and pass the requests on to port 80 in Traefik.

raykrueger commented:
@krancour Thanks for the input, but I'm having a hard time finding action you want me to take. My patch keeps the same functionality as before by terminating SSL at the service and forwarding to a nonsecure port internally. Due to the port conflict it just can't be 80.

raykrueger committed:
Add the new containerPort 84

shadycuz commented:
Also having this problem on 1.8.1 with default values.

krancour commented:
You're right. Let me re-review in the morning.

krancour commented:
@raykrueger thanks for pushing back on me. I read the changes wrong the first time through. I agree with what you're doing here. This LGTM.

If I can make one suggestion, it would be to use an unprivileged port that is frequently associated with unsecured HTTP traffic (like 8080) instead of using the privileged port of 84. I checked and IANA hasn't designated 84 for anything, but I think that in general, when using a port for any purpose that it's not commonly associated with, it's better to use an unprivileged one.

Whether you take that suggestion or not, this LGTM either way.

raykrueger commented:
I can make that change, sure. I was also going to ask about turning the http service port off when https is turned on. A sort of “why secure the front door if we’re going to leave the back door open” type question.

krancour commented:
@raykrueger

> I was also going to ask about turning the http service port off when https is turned on.

Good question. _Now_ the answer I first gave is the correct answer. It's because of people who want to terminate SSL at the load balancer.

This chart supports adding arbitrary annotations to the service and if I'm not mistaken, on some cloud providers, there are annotations that can be used to manage SSL termination. Even for cloud providers that have no such option, it's possible to manually modify the load balancer that the cloud provider creates for you to ensure termination there.

In such cases, you want the secured front door into the load balancer connected to the unsecured door into the Traefik pod(s).

hlascelles commented:
Thanks for looking at this, it's a pain for us too.

The Kubernetes team [are in the process](https://github.com/kubernetes/kubernetes/issues/47222) of reverting the breaking change and have just [opened a PR to do so](https://github.com/kubernetes/kubernetes/pull/53576), so it looks like this problem will go away (hopefully).

My go-templating-fu is quite fresh. After these changes, is there a way to have it so we can get http->https redirect for all requests (handled by traefik), then https(443)->ELBSSLtermination->http(80) for all other requests? We're using an AWS ELB with AMC.

It /appears/ like `ssl.enabled` must be on for us to get `entryPoints.http.redirect`. We don't want `ssl.enabled` as we don't need certs applied in traefik. Have I understood it right?

raykrueger commented:
@krancour Sorry, yes. I wasn't thinking of external loadbalancers vs the Kubernetes service. I gotcha. I'll make the 8080 change now too.

raykrueger committed:
Switch the alternate http port to 8080

After discussing the issue on github I'm moving the alternate http port
from 84 to 8080.

raykrueger commented:
@krancour Actually, I should have noticed this first, there's the dashboard on 8080.

raykrueger committed:
Use port 8880 to avoid conflicts. Use named ports

In another chart I worked on it was recommended to use named container
ports. I'm following that advice here.

I am also avoiding the 8080 dashboard conflict by switching the http
alternate port (httpn) to 8880.

raykrueger commented:
I've manually tested all the ports. Everything appears to work.

krancour commented:
LGTM

cc @prydonius @viglesiasce

krancour commented:
@hlascelles 

> It /appears/ like ssl.enabled must be on for us to get entryPoints.http.redirect. We don't want ssl.enabled as we don't need certs applied in traefik. Have I understood it right?

If redirects cannot be enabled without also enabling SSL on Traefik, that sounds like a shortcoming of Traefik itself. You may want to open a feature request.

hlascelles commented:
I believe it's a limitation of this chart. See [this line](https://github.com/kubernetes/charts/blob/master/stable/traefik/templates/configmap.yaml#L26)

The chart is assuming that ssl can only be enforced if ssl is enabled _in traefik_. In our case, ssl is "enabled" elsewhere.

It looks like it's a different change to this PR, so shall I create a new one? The proposal is to change:

`{{- if and .Values.ssl.enabled .Values.ssl.enforced }}`
to
`{{- if .Values.ssl.enforced }}`
